### PR TITLE
Shrink the image size and reduce layers

### PR DESCRIPTION
**Pull request by fsimonis**
Proxy for #39

### CODE REVIEW

File: Dockerfile.Ubuntu1604 [hunk 1]
```diff
@@ -14,38 +14,44 @@ RUN apt-get -qq update && apt-get -qq install \
...
 # Some parameters for the build, you can set them in the build command e.g.
 # sudo docker build Dockerfile.precice --build-arg petsc_para=yes --build-arg mpi_para=yes .
 # this will result in
 # scons petsc=yes mpi=yes python=no compiler="mpicxx" -j2
 ARG petsc_para=no
 ARG mpi_para=yes
 ARG python_para=no
-RUN scons petsc=$petsc_para mpi=$mpi_para python=$python_para -j$(nproc)
+RUN scons petsc=$petsc_para mpi=$mpi_para python=$python_para -j$(nproc) && \
+    find . -type f \( -name "*.o" -or -name "*.os" \) -exec rm {} +
```
Comment (floli) [inline Dockerfile.Ubuntu1604:51]: Please add a comment, why this is necessary. Don't think it's obvious that's just for reducing image size.

File: Dockerfile.Ubuntu1604 [hunk 2]
```diff
@@ -14,38 +14,44 @@ RUN apt-get -qq update && apt-get -qq install \
     python-numpy \
     python-dev \
     wget \
-    bzip2
+    bzip2 && \
+    rm -rf /var/lib/apt/lists/*
```
Comment (MakisH) [inline Dockerfile.Ubuntu1604:18]: Why not do a normal `apt-get clean`?

File: Dockerfile.Ubuntu1804 [hunk 3]
```diff
@@ -15,31 +15,35 @@ RUN apt-get -qq update && apt-get -qq install \
     python-numpy \
     python-dev \
     wget \
-    bzip2    
+    bzip2 && \
+    rm -rf /var/lib/apt/lists/*
```
Comment (MakisH) [inline Dockerfile.Ubuntu1804:19]: Same here about `apt-get clean`.

File: Test_bindings.Ubuntu1804/Dockerfile [hunk 4]
```diff
@@ -9,7 +9,8 @@ ENV USER=root
 
 # Installing necessary dependecies
 RUN apt-get -qq update && apt-get -qq install \
-    cython python-numpy python-enum34 python-mpi4py cython3 python3-numpy python3-mpi4py
+    cython python-numpy python-enum34 python-mpi4py cython3 python3-numpy python3-mpi4py && \
+    rm -rf /var/lib/apt/lists/*
```
Comment (MakisH) [inline Test_bindings.Ubuntu1804/Dockerfile:13]: Same here about `apt-get clean`.

File: Test_bindings/Dockerfile [hunk 5]
```diff
@@ -9,13 +9,14 @@ ENV USER=root
 
 # Installing necessary dependecies
 RUN apt-get -qq update && apt-get -qq install \
-    python-pip python-enum34 python3-pip
+    python-pip python-enum34 python3-pip && \
+    rm -rf /var/lib/apt/lists/*
```
Comment (MakisH) [inline Test_bindings/Dockerfile:12]: Well, same everywhere (`apt-get clean`).

**Review by MakisH — CHANGES_REQUESTED** [review]
My only requested change is using `apt-get clean` instead of manually deleting files.

Apart from this, we could use [Apline Linux](https://coderwall.com/p/s4ofoq/docker-why-i-prefer-alpine-as-base-instead-of-ubuntu) instead of Ubuntu to really reduce the image size. But then we would not reproduce an Ubuntu testing environment, of course. However, it could be useful if one day we wanted to ship Docker images.

File: Dockerfile.Ubuntu1604 (hunk 1 above)
Comment (fsimonis) [reply Dockerfile.Ubuntu1604:51]: True, alternatively we could wipe the entire directory apart from the library and executables.

File: Dockerfile.Ubuntu1604 (hunk 2 above)
Comment (fsimonis) [reply Dockerfile.Ubuntu1604:18]: Ubuntu Docker images run `apt-get clean` after each install. [More here](https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#minimize-the-number-of-layers)

File: Dockerfile.Ubuntu1604 (hunk 1 above)
Comment (fsimonis) [reply Dockerfile.Ubuntu1604:51]: I added a comment regarding the clean-up.
Feel free to resolve the conversation if you are ok with it.

**Comment by fsimonis** [comment]
@MakisH I don't think Alpine is a good choice here. We try to model a plausible environment, so the Ubuntu images may be the best choice.

**Comment by MakisH** [comment]
Interesting piece of information, for any other projects trying to reduce their image size as well:

> In Docker 1.10 and higher, only the instructions RUN, COPY, ADD create layers. Other instructions create temporary intermediate images, and do not directly increase the size of the build.

So, at the end, merging the `ENV` may not have been necessary.

From the (linked above by @fsimonis ) [Docker best practices for writing Dockerfiles](https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#minimize-the-number-of-layers).

**Edit:** actually, this is reverted further below:

> Each ENV line creates a new intermediate layer, just like RUN commands. This means that even if you unset the environment variable in a future layer, it still persists in this layer and its value can be dumped.

**Comment by MakisH** [comment]
Travis error:

```
Traceback (most recent call last):
  File "/usr/bin/pip3", line 9, in <module>
    from pip import main
ImportError: cannot import name 'main'
The command '/bin/sh -c pip2 install --upgrade pip &&     pip2 install Cython mpi4py &&     pip3 install --upgrade pip &&     pip3 install Cython mpi4py numpy' returned a non-zero code: 1
```

**Comment by fsimonis** [comment]
@MakisH I merged the ENV to prevent the generation of the intermediate layers. My hope is that docker has less to do this way.

**Review by shkodm — APPROVED** [review]

**Review by MakisH — APPROVED** [review]

**Pull request merged**